rest/route/pod_agent.go

```diff
@@ -328,3 +348,155 @@ func (h *podAgentNextTask) findDispatcher() (*dispatcher.PodDispatcher, error) {
 
 	return pd, nil
 }
+
+// POST /rest/v2/pods/{pod_id}/task/{task_id}/end
+
+type podAgentEndTask struct {
+	env     evergreen.Environment
+	podID   string
+	taskID  string
+	details apimodels.TaskEndDetail
+}
+
+func makePodAgentEndTask(env evergreen.Environment) gimlet.RouteHandler {
+	return &podAgentEndTask{
+		env: env,
+	}
+}
+
+func (h *podAgentEndTask) Factory() gimlet.RouteHandler {
+	return &podAgentEndTask{
+		env: h.env,
+	}
+}
+
+func (h *podAgentEndTask) Parse(ctx context.Context, r *http.Request) error {
+	h.podID = gimlet.GetVars(r)["pod_id"]
+	if h.podID == "" {
+		return errors.New("missing pod ID")
+	}
+	h.taskID = gimlet.GetVars(r)["task_id"]
+	if h.taskID == "" {
+		return errors.New("missing task ID")
+	}
+	details := apimodels.TaskEndDetail{}
+	if err := utility.ReadJSON(utility.NewRequestReader(r), &details); err != nil {
+		return errors.Wrap(err, "reading task end details from JSON request body")
+	}
+	h.details = details
+
+	// Check that status is either finished or aborted (i.e. undispatched)
+	if !evergreen.IsValidTaskEndStatus(details.Status) && details.Status != evergreen.TaskUndispatched {
+		return errors.Errorf("invalid end status '%s' for task %s", details.Status, h.taskID)
+	}
+	return nil
+}
+
+// Run finishes the given task and generates a job to collect task end stats.
+// It then marks the task as finished or inactive if aborted.
+func (h *podAgentEndTask) Run(ctx context.Context) gimlet.Responder {
+	finishTime := time.Now()
+	p, err := findPod(h.podID)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(err)
+	}
+	t, err := findTask(h.taskID)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(err)
+	}
+
+	endTaskResp := &apimodels.EndTaskResponse{}
+
+	if p.RunningTask == "" {
+		grip.Notice(message.Fields{
+			"message":                 "pod is not assigned task, not clearing, asking agent to exit",
+			"task_id":                 t.Id,
+			"task_status_from_db":     t.Status,
+			"task_details_from_db":    t.Details,
+			"current_agent":           p.AgentVersion == evergreen.AgentVersion,
+			"agent_version":           p.AgentVersion,
+			"build_revision":          evergreen.BuildRevision,
+			"build_agent":             evergreen.AgentVersion,
+			"task_details_from_agent": h.details,
+			"pod_id":                  p.ID,
+		})
+		endTaskResp.ShouldExit = true
+		return gimlet.NewJSONResponse(endTaskResp)
+	}
+
+	// Clear the running task on the pod now that the task has finished.
+	if err := p.ClearRunningTask(); err != nil {
```

Comment:
Is this safe to do before `model.MarkEnd` has succeeded? I'm concerned that if something happens in between this line and when `model.MarkEnd` actually runs and succeeds, we might tell the pod that it's done running this task and it'll be allowed to move on before the task, abandoning this task in a stuck and incomplete state.